bftengine/src/bftengine/TimeServiceManager.hpp

```diff
@@ -0,0 +1,98 @@
+    if (new_time > last_timestamp) {
+      client_.setLastTimestamp(new_time);
+      return new_time;
+    }
+
+    LOG_INFO(TS_MNGR,
+             "New time(" << new_time.count() << "ms since epoch) is less or equal to reserved ("
+                         << last_timestamp.count() << "), new time will be "
+                         << (last_timestamp + config_.timeServiceEpsilonMillis).count());
+    last_timestamp += config_.timeServiceEpsilonMillis;
+    client_.setLastTimestamp(last_timestamp);
+    return last_timestamp;
+  }
+
+  // Returns a client request message with timestamp (current system clock time)
+  [[nodiscard]] std::unique_ptr<impl::ClientRequestMsg> createClientRequestMsg() {
+    const auto now = std::chrono::duration_cast<ConsensusTime>(ClockT::now().time_since_epoch());
+    const auto& serialized = concord::util::serialize(now);
+    return std::make_unique<impl::ClientRequestMsg>(config_.replicaId,
+                                                    MsgFlag::TIME_SERVICE_FLAG,
+                                                    0U,
+                                                    serialized.size(),
+                                                    serialized.data(),
+                                                    std::numeric_limits<uint64_t>::max());
+  }
+
+  [[nodiscard]] bool isPrimarysTimeWithinBounds(impl::ClientRequestMsg& msg) const {
+    ConcordAssert((msg.flags() & MsgFlag::TIME_SERVICE_FLAG) != 0 &&
+                  "TimeServiceManager supports only messages with TIME_SERVICE_FLAG");
+    const auto t = concord::util::deserialize<ConsensusTime>(msg.requestBuf(), msg.requestBuf() + msg.requestLength());
+    const auto now = std::chrono::duration_cast<ConsensusTime>(ClockT::now().time_since_epoch());
+
+    auto min = now - config_.timeServiceHardLimitMillis;
+    auto max = now + config_.timeServiceHardLimitMillis;
+    if (min > t || t > max) {
+      // TODO(DD): Add metrics
+      LOG_ERROR(TS_MNGR,
+                "Current primary's time reached hard limit, requests will be ignored. Please synchronize local clocks! "
+                    << "Primary's time: " << t.count() << ", local time: " << now.count() << ", time limits: +/-"
+                    << config_.timeServiceHardLimitMillis.count() << ". Time is presented as ms since epoch");
+      return false;
+    }
+
+    min = now - config_.timeServiceSoftLimitMillis;
+    max = now + config_.timeServiceSoftLimitMillis;
+    if (min > t || t > max) {
+      // TODO(DD): Add metrics
+      LOG_WARN(TS_MNGR,
+               "Current primary's time reached soft limit, please synchronize local clocks! "
+                   << "Primary's time: " << t.count() << ", local time: " << now.count() << ", time limits: +/-"
+                   << config_.timeServiceSoftLimitMillis.count() << ". Time is presented as ms since epoch");
+    }
+    return true;
+  }
+
+ private:
+  TimeServiceResPageClient client_;
+  const ReplicaConfig& config_;
```

Comment:
ReplicaConfig is a singleton now, no need to pass.

Reply:
Done